File: messageformat/js/src/mf1.ts
/**
 * ICU MessageFormat 1 → canonical MF2 data model.
 *
 * Built on the reference conversion (`mf1ToMessageData`), then canonicalized
 * so MF1 lands on *standard* MF2 functions wherever one is equivalent:
 *
 * | MF1                                    | MF2                                        |
 * |----------------------------------------|--------------------------------------------|
 * | `{n, number}` / `integer`              | `:number` / `:integer`                     |
 * | `{n, number, percent}`, `::percent scale/100` | `:percent`                          |
 * | `{n, number, ::currency/EUR}`          | `:currency currency=EUR`                   |
 * | `{n, number, ::measure-unit/…}`        | `:unit unit=…`                             |
 * | `{d, date[, short/medium/long/full]}`  | `:date [length=…] [fields=…]`              |
 * | `{t, time[, short/medium/long/full]}`  | `:time precision=… [timeZoneStyle=…]`      |
 * | `select`                               | `:string` selector                         |
 * | `plural` / `selectordinal`             | `:number` / `:number select=ordinal`       |
 * | `plural, offset:N`                     | `:number` + `.local $x_offset = {$x :offset subtract=N}` |
 *
 * Everything else keeps an `mf1:`-namespaced function (`mf1:currency` without
 * a currency code, `mf1:number` for notations/scales, `mf1:date` for
 * skeletons, `spellout`, `duration`, …). The original argument type and style
 * stay on the expression as `@mf1:argType` / `@mf1:argStyle` attributes, so
 * the message can be exported back to MF1.
 */
import { mf1ToMessageData } from "@messageformat/icu-messageformat-1";
import { parse } from "@messageformat/parser";
import type { PluralCategory, Token } from "@messageformat/parser";
import { fromReference } from "./convert.js";
import type {
  Declaration,
  Expression,
  FunctionRef,
  Message,
  Pattern,
  SelectMessage,
} from "./model.js";

/** A syntax error (or a plural key the locale never uses) in an MF1 message. */
export class MF1SyntaxError extends Error {
  override name = "MF1SyntaxError";
}

type Tokens = Token[];

/**
 * Parse an ICU MessageFormat 1 message into the canonical data model.
 * `locale` decides which plural and ordinal keys are valid.
 *
 * @throws MF1SyntaxError
 */
export function parseMF1(src: string, locale: string): Message {
  let ast: Tokens;
  try {
    ast = parse(src, {
      cardinal: categories(locale, "cardinal"),
      ordinal: categories(locale, "ordinal"),
    });
  } catch (error) {
    throw new MF1SyntaxError(error instanceof Error ? error.message : String(error), {
      cause: error,
    });
  }
  const offsetNames = offsetVariables(ast);
  const msg = fromReference(mf1ToMessageData(ast as Parameters<typeof mf1ToMessageData>[0]));
  return canonicalize(msg, offsetNames);
}

function categories(locale: string, type: Intl.PluralRuleType): PluralCategory[] {
  return new Intl.PluralRules(locale, { type }).resolvedOptions()
    .pluralCategories as PluralCategory[];
}

// ── plural offsets ────────────────────────────────────────────────────

const isSelect = (t: Token) =>
  t.type === "plural" || t.type === "select" || t.type === "selectordinal";

function argNames(tokens: Tokens, names = new Set<string>()): Set<string> {
  for (const t of tokens) {
    if (t.type === "argument" || t.type === "function") names.add(t.arg);
    if (t.type === "function" && t.param) argNames(t.param, names);
    if (isSelect(t) && "cases" in t) {
      names.add(t.arg);
      for (const c of t.cases) argNames(c.tokens, names);
    }
  }
  return names;
}

/**
 * Name the offset variable of every `plural` with an offset, and point its
 * `#` at that variable: in MF1, `#` shows the value minus the offset, while
 * `{arg}` keeps showing the raw value.
 */
function offsetVariables(ast: Tokens): Map<string, string> {
  const taken = argNames(ast);
  const names = new Map<string, string>();
  const nameFor = (arg: string) => {
    let name = names.get(arg);
    if (!name) {
      name = `${arg}_offset`;
      while (taken.has(name)) name += "_";
      taken.add(name);
      names.set(arg, name);
    }
    return name;
  };
  const walk = (tokens: Tokens, octothorpe: string | null) => {
    tokens.forEach((t, i) => {
      if (t.type === "octothorpe" && octothorpe) {
        tokens[i] = { type: "argument", arg: octothorpe, ctx: t.ctx };
      } else if (isSelect(t) && "cases" in t) {
        const inner =
          t.type === "plural" && t.pluralOffset
            ? nameFor(t.arg)
            : t.type === "select"
              ? octothorpe
              : null;
        for (const c of t.cases) walk(c.tokens, inner);
      }
    });
  };
  walk(ast, null);
  return names;
}

// ── canonicalization ──────────────────────────────────────────────────

function canonicalize(msg: Message, offsetNames: Map<string, string>): Message {
  for (const d of msg.declarations) d.value = canonicalExpression(d.value) as typeof d.value;
  if (msg.type === "message") {
    msg.pattern = canonicalPattern(msg.pattern);
    return msg;
  }
  for (const v of msg.variants) v.value = canonicalPattern(v.value);
  for (const d of [...msg.declarations]) {
    const offset = pluralOffset(d);
    const name = offsetNames.get(d.name);
    if (offset !== undefined && name) splitOffsetSelector(msg, d, offset, name);
  }
  return msg;
}

const canonicalPattern = (p: Pattern): Pattern =>
  p.map((el) =>
    typeof el !== "string" && el.type === "expression" ? canonicalExpression(el) : el,
  );

function canonicalExpression(expr: Expression): Expression {
  if (expr.function) expr.function = canonicalFunction(expr.function);
  return expr;
}

function canonicalFunction(fn: FunctionRef): FunctionRef {
  const opts = { ...fn.options };
  const lit = (name: string) => {
    const o = opts[name];
    return o?.type === "literal" ? o.value : undefined;
  };
  const has = (name: string) => name in opts;
  const rename = (name: string, drop: string[] = []): FunctionRef => {
    for (const d of drop) delete opts[d];
    const out: FunctionRef = { type: "function", name };
    if (Object.keys(opts).length) out.options = opts;
    return out;
  };
  switch (fn.name) {
    case "mf1:unit":
      if (lit("mf1:scale") === "100" && (!has("unit") || lit("unit") === "percent")) {
        return rename("percent", ["unit", "mf1:scale"]);
      }
      return has("unit") && !has("mf1:scale") ? rename("unit") : fn;
    case "mf1:currency":
      return has("currency") && !has("mf1:scale") ? rename("currency") : fn;
    case "mf1:date":
      return has("mf1:argStyle") ? fn : rename("date");
    case "mf1:time":
      if (has("mf1:argStyle")) return fn;
      if (opts.timeZoneName) {
        opts.timeZoneStyle = opts.timeZoneName;
        delete opts.timeZoneName;
      }
      return rename("time");
    default:
      return fn;
  }
}

/** The offset of an `.input {$x :mf1:plural offset=N}` cardinal declaration. */
function pluralOffset(d: Declaration): number | undefined {
  const fn = d.value.function;
  if (d.type !== "input" || fn?.name !== "mf1:plural") return undefined;
  const offset = fn.options?.offset;
  if (fn.options?.select || offset?.type !== "literal" || !/^\d+$/.test(offset.value))
    return undefined;
  return Number(offset.value);
}

/**
 * `.input {$c :mf1:plural offset=1} .match $c` becomes
 * `.input {$c :number} .local $c_offset = {$c :offset subtract=1} .match $c $c_offset`:
 * exact keys (`=0`) stay on `$c`, plural categories move to `$c_offset`.
 */
function splitOffsetSelector(
  msg: SelectMessage,
  d: Declaration,
  offset: number,
  name: string,
): void {
  const arg = { type: "variable", name: d.name } as const;
  d.value = { ...d.value, function: { type: "function", name: "number" } };
  msg.declarations.push({
    type: "local",
    name,
    value: {
      type: "expression",
      arg,
      function: {
        type: "function",
        name: "offset",
        options: { subtract: { type: "literal", value: String(offset) } },
      },
    },
  });
  const i = msg.selectors.findIndex((s) => s.name === d.name);
  if (i < 0) return;
  msg.selectors.splice(i + 1, 0, { type: "variable", name });
  for (const v of msg.variants) {
    const key = v.keys[i];
    if (!key) continue;
    const exact = key.type === "literal" && /^\d+$/.test(key.value);
    v.keys.splice(i, 1, exact ? key : { type: "*" }, exact ? { type: "*" } : key);
  }
}

File: messageformat/mf1.go
package messageformat

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/felixgeelhaar/glossa/messageformat/internal/cldr"
	"github.com/kaptinlin/messageformat-go/mf1"
	"golang.org/x/text/language"
)

// ICU MessageFormat 1 -> canonical MF2 conversion.
//
// The approach follows mf1ToMessageData of @messageformat/icu-messageformat-1
// (nested selects are lifted into one top-level .match), with one Glossa
// decision: MF1 constructs map onto standard MF2 functions wherever those
// reproduce the MF1 output, so runtimes only need the standard function
// set. testdata/glossa/README.md lists the rules; testdata/glossa/
// mf1-to-mf2.json proves them against the reference implementations.

// maxVariants bounds the cartesian product of lifted selectors, so a
// pathological message cannot exhaust memory.
const maxVariants = 4096

// ParseMF1 parses ICU MessageFormat 1 source and converts it to the
// canonical MF2 model. locale (a BCP 47 tag) decides which plural keys are
// valid, as in the MF1 reference implementation.
//
// Errors: CodeInvalidLocale, CodeMF1SyntaxError for source that does not
// parse (including plural keys the locale doesn't have), and
// CodeMF1Unsupported for valid MF1 that has no MF2 representation.
func ParseMF1(src, locale string) (msg Message, err error) {
	defer containEngineFailure(&err)
	return parseMF1(src, locale)
}

func parseMF1(src, locale string) (Message, error) {
	opts, err := mf1ParseOptions(locale)
	if err != nil {
		return Message{}, err
	}
	tokens, err := mf1.Parse(src, opts)
	if err != nil {
		return Message{}, &Error{Code: CodeMF1SyntaxError, Message: err.Error()}
	}
	msg, err := newMF1Converter(tokens).convert()
	if err != nil {
		return Message{}, &Error{Code: CodeMF1Unsupported, Message: err.Error()}
	}
	if err := Validate(msg); err != nil {
		return Message{}, &Error{Code: CodeMF1Unsupported, Message: fmt.Sprintf("converted message is not valid MF2: %v", err)}
	}
	return msg, nil
}

func mf1ParseOptions(locale string) (*mf1.ParseOptions, error) {
	if _, err := language.Parse(locale); err != nil {
		return nil, &Error{Code: CodeInvalidLocale, Message: fmt.Sprintf("%q: %v", locale, err)}
	}
	cardinal, err := cldr.PluralCategories(locale, false)
	if err != nil {
		return nil, &Error{Code: CodeInvalidLocale, Message: err.Error()}
	}
	ordinal, err := cldr.PluralCategories(locale, true)
	if err != nil {
		return nil, &Error{Code: CodeInvalidLocale, Message: err.Error()}
	}
	return &mf1.ParseOptions{Cardinal: asMF1Categories(cardinal), Ordinal: asMF1Categories(ordinal)}, nil
}

func asMF1Categories(cats []string) []mf1.PluralCategory {
	out := make([]mf1.PluralCategory, len(cats))
	for i, c := range cats {
		out[i] = mf1.PluralCategory(c)
	}
	return out
}

// --- selector analysis --------------------------------------------------------

type mf1SelectKind string

const (
	mf1Select        mf1SelectKind = "select"
	mf1Plural        mf1SelectKind = "plural"
	mf1SelectOrdinal mf1SelectKind = "selectordinal"
)

// mf1Selector is one MF1 argument used as a selector, with every key it is
// matched against anywhere in the message.
type mf1Selector struct {
	name   string
	kind   mf1SelectKind
	offset int
	exact  []string // plural `=n` keys as MF2 numeric literals
	keys   []string // select keys or plural categories, without `other`

	displayVar string // what `#` shows: the variable, or its :offset local
	exactCol   int    // column of exact keys (the variable itself), or -1
	keyCol     int    // column of the other keys, or -1
}

func (s *mf1Selector) isPlural() bool { return s.kind != mf1Select }

type mf1Converter struct {
	tokens    []mf1.Token
	selectors []*mf1Selector
	byName    map[string]*mf1Selector
	names     map[string]bool // every variable name in the message
	columns   []column
	variants  []Variant
}

// column is one MF2 selector of the lifted .match.
type column struct {
	variable string
	keys     []string // literal keys in order; the catch-all is implied last
}

func newMF1Converter(tokens []mf1.Token) *mf1Converter {
	return &mf1Converter{tokens: tokens, byName: map[string]*mf1Selector{}, names: map[string]bool{}}
}

func (c *mf1Converter) convert() (Message, error) {
	if err := c.collect(c.tokens); err != nil {
		return Message{}, err
	}
	if len(c.selectors) == 0 {
		pattern := Pattern{}
		for _, tok := range c.tokens {
			el, err := c.leaf(tok, nil)
			if err != nil {
				return Message{}, err
			}
			pattern = appendElement(pattern, el)
		}
		return Message{Type: PatternMessageType, Pattern: pattern}, nil
	}
	decls := c.planColumns()
	if err := c.buildVariants(); err != nil {
		return Message{}, err
	}
	if err := c.addParts(c.tokens, nil, nil); err != nil {
		return Message{}, err
	}
	selectors := make([]VariableRef, len(c.columns))
	for i, col := range c.columns {
		selectors[i] = VariableRef{Name: col.variable}
	}
	return Message{Type: SelectMessageType, Declarations: decls, Selectors: selectors, Variants: c.variants}, nil
}

// collect records every selector (depth first, in source order) and every
// variable name.
func (c *mf1Converter) collect(tokens []mf1.Token) error {
	for _, tok := range tokens {
		switch t := tok.(type) {
		case *mf1.PlainArg:
			c.names[t.Arg] = true
		case *mf1.FunctionArg:
			c.names[t.Arg] = true
		case *mf1.Select:
			c.names[t.Arg] = true
			if err := c.addSelector(t); err != nil {
				return err
			}
			for _, sc := range t.Cases {
				if err := c.collect(sc.Tokens); err != nil {
					return err
				}
			}
		}
	}
	return nil
}

func (c *mf1Converter) addSelector(t *mf1.Select) error {
	kind := mf1SelectKind(t.Type)
	offset := 0
	if t.PluralOffset != nil {
		offset = *t.PluralOffset
	}
	sel, seen := c.byName[t.Arg]
	if !seen {
		sel = &mf1Selector{name: t.Arg, kind: kind, offset: offset, exactCol: -1, keyCol: -1}
		c.byName[t.Arg] = sel
		c.selectors = append(c.selectors, sel)
	} else if sel.kind != kind || sel.offset != offset {
		return fmt.Errorf("argument %q is used both as %s and as %s (offset %d vs %d); MF2 needs one declaration per variable",
			t.Arg, sel.kind, kind, sel.offset, offset)
	}
	for _, sc := range t.Cases {
		switch {
		case sc.Key == "other":
		case sel.isPlural() && strings.HasPrefix(sc.Key, "="):
			sel.exact = appendUnique(sel.exact, normalizeExactKey(sc.Key[1:]))
		default:
			sel.keys = appendUnique(sel.keys, sc.Key)
		}
	}
	return nil
}

// normalizeExactKey turns the digits of an `=n` key into the MF2 numeric
// literal that :number matches exactly (`=01` -> `1`).
func normalizeExactKey(digits string) string {
	if n, err := strconv.Atoi(digits); err == nil {
		return strconv.Itoa(n)
	}
	return digits
}

// planColumns assigns MF2 selector columns and returns the declarations.
func (c *mf1Converter) planColumns() []Declaration {
	var decls []Declaration
	for _, s := range c.selectors {
		decls = append(decls, inputDeclaration(s))
		s.displayVar = s.name
		if !s.isPlural() || s.offset == 0 {
			s.keyCol = c.addColumn(s.name, append(append([]string{}, s.exact...), s.keys...))
			s.exactCol = s.keyCol
			continue
		}
		local := c.uniqueName(fmt.Sprintf("%s_minus_%d", s.name, s.offset))
		s.displayVar = local
		decls = append(decls, Declaration{
			Type: LocalDeclaration,
			Name: local,
			Value: Expression{
				Arg:      VariableRef{Name: s.name},
				Function: &FunctionRef{Name: "offset", Options: Options{"subtract": Literal{Value: strconv.Itoa(s.offset)}}},
			},
		})
		if len(s.exact) > 0 {
			s.exactCol = c.addColumn(s.name, s.exact)
		}
		if len(s.keys) > 0 || len(s.exact) == 0 {
			s.keyCol = c.addColumn(local, s.keys)
		}
	}
	return decls
}

func inputDeclaration(s *mf1Selector) Declaration {
	fn := &FunctionRef{Name: "number"}
	switch s.kind {
	case mf1Select:
		fn = &FunctionRef{Name: "string"}
	case mf1SelectOrdinal:
		fn.Options = Options{"select": Literal{Value: "ordinal"}}
	}
	return Declaration{Type: InputDeclaration, Name: s.name, Value: Expression{Arg: VariableRef{Name: s.name}, Function: fn}}
}

func (c *mf1Converter) addColumn(variable string, keys []string) int {
	c.columns = append(c.columns, column{variable: variable, keys: keys})
	return len(c.columns) - 1
}

func (c *mf1Converter) uniqueName(base string) string {
	name := base
	for i := 2; c.names[name]; i++ {
		name = fmt.Sprintf("%s_%d", base, i)
	}
	c.names[name] = true
	return name
}

// buildVariants creates the cartesian product of all column keys, first
// column outermost, each column's catch-all last.
func (c *mf1Converter) buildVariants() error {
	total := 1
	for _, col := range c.columns {
		total *= len(col.keys) + 1
		if total > maxVariants {
			return fmt.Errorf("the lifted message would need more than %d variants", maxVariants)
		}
	}
	keys := [][]VariantKey{{}}
	for _, col := range c.columns {
		options := make([]VariantKey, 0, len(col.keys)+1)
		for _, k := range col.keys {
			options = append(options, VariantKey{Value: k})
		}
		options = append(options, VariantKey{Catchall: true})
		next := make([][]VariantKey, 0, len(keys)*len(options))
		for _, prefix := range keys {
			for _, k := range options {
				next = append(next, append(append([]VariantKey{}, prefix...), k))
			}
		}
		keys = next
	}
	c.variants = make([]Variant, len(keys))
	for i, k := range keys {
		c.variants[i] = Variant{Keys: k, Value: Pattern{}}
	}
	return nil
}

// --- pattern construction -----------------------------------------------------

// constraint restricts the variants a token belongs to: column col must
// have key (the empty string for the catch-all).
type constraint struct {
	col int
	key string
}

// addParts appends every leaf token to each variant that satisfies filter.
// plural is the innermost enclosing plural, which `#` refers to.
func (c *mf1Converter) addParts(tokens []mf1.Token, plural *mf1Selector, filter []constraint) error {
	for _, tok := range tokens {
		sel, ok := tok.(*mf1.Select)
		if !ok {
			el, err := c.leaf(tok, plural)
			if err != nil {
				return err
			}
			c.appendMatching(el, filter)
			continue
		}
		s := c.byName[sel.Arg]
		inner := plural
		if s.isPlural() {
			inner = s
		}
		for _, sc := range sel.Cases {
			cons := append(append([]constraint{}, filter...), caseConstraints(s, sc.Key)...)
			if err := c.addParts(sc.Tokens, inner, cons); err != nil {
				return err
			}
		}
	}
	return nil
}

// caseConstraints maps one MF1 case key to column constraints.
func caseConstraints(s *mf1Selector, key string) []constraint {
	switch {
	case s.isPlural() && strings.HasPrefix(key, "="):
		return []constraint{{s.exactCol, normalizeExactKey(key[1:])}}
	case key == "other":
		var out []constraint
		if s.exactCol >= 0 {
			out = append(out, constraint{s.exactCol, ""})
		}
		if s.keyCol >= 0 && s.keyCol != s.exactCol {
			out = append(out, constraint{s.keyCol, ""})
		}
		return out
	default:
		out := []constraint{{s.keyCol, key}}
		if s.exactCol >= 0 && s.exactCol != s.keyCol {
			out = append(out, constraint{s.exactCol, ""})
		}
		return out
	}
}

func (c *mf1Converter) appendMatching(el PatternElement, filter []constraint) {
	for i := range c.variants {
		v := &c.variants[i]
		if variantMatches(v.Keys, filter) {
			v.Value = appendElement(v.Value, clonePatternElement(el))
		}
	}
}

func variantMatches(keys []VariantKey, filter []constraint) bool {
	for _, f := range filter {
		k := keys[f.col]
		if f.key == "" {
			if !k.Catchall {
				return false
			}
		} else if k.Catchall || k.Value != f.key {
			return false
		}
	}
	return true
}

// leaf converts a non-select token to a pattern element.
func (c *mf1Converter) leaf(tok mf1.Token, plural *mf1Selector) (PatternElement, error) {
	switch t := tok.(type) {
	case *mf1.Content:
		return Text(t.Value), nil
	case *mf1.PlainArg:
		return Expression{Arg: VariableRef{Name: t.Arg}}, nil
	case *mf1.FunctionArg:
		return convertFunctionArg(t)
	case *mf1.Octothorpe:
		if plural == nil {
			return Text("#"), nil
		}
		return Expression{Arg: VariableRef{Name: plural.displayVar}}, nil
	default:
		return nil, fmt.Errorf("unsupported MF1 token %q", tok.GetType())
	}
}

// appendElement appends el to p, merging adjacent text.
func appendElement(p Pattern, el PatternElement) Pattern {
	if text, ok := el.(Text); ok && len(p) > 0 {
		if last, ok := p[len(p)-1].(Text); ok {
			p[len(p)-1] = last + text
			return p
		}
	}
	return append(p, el)
}

func appendUnique(list []string, s string) []string {
	for _, v := range list {
		if v == s {
			return list
		}
	}
	return append(list, s)
}
